singer-io/tap-linkedin-ads
# Sync all fields

# Description of change
LinkedIn made a change to their API to limit the number of fields we can sync. This PR allows you to sync all fields for the `ad_analytics_by_campaign` stream.

The current implementation of `sync_endpoint()` has 3 jobs:
- Get a page of records from the API
- `transform_json()`
- `process_records()`

This PR breaks `ad_analytics_by_campaign` out of that flow, but tries to mirror it as closely as possible.

The only change is to the "Get a page of records from the API" step. Depending on the number of fields selected, we make multiple requests to get the data and then merge these requests into the format that `transform_json()` expects.

# Manual QA steps
- See #17 
- Ran the tap selecting the same fields as the last test in #17, ensured this version produced the same records
- Ran the tap selecting all fields, ensured the fields from the previous sync look the same
- There are some unit tests for the new helper functions

# Risks
- Low because we compared the outputted records to what the previous version of the tap produced
- One thing came up in testing: `Both syncs saw cost_in_local_currency but got different values 49.02123 | 49.02`
  - Rounding is out of our control here
- More concerning is that a more recent sync saw a lower number: `Both syncs saw approximate_unique_impressions but got different values 3639 | 3638`
 
# Rollback steps
 - revert this branch
